packages/artifact/src/internal/upload-http-client.ts

```diff
@@ -208,25 +208,30 @@ export class UploadHttpClient {
     // for creating a new GZip file, an in-memory buffer is used for compression
     if (totalFileSize < 65536) {
       const buffer = await createGZipFileInBuffer(parameters.file)
-      let uploadStream: NodeJS.ReadableStream
+
+      //An open stream is needed in the event of a failure and we need to retry. If a NodeJS.ReadableStream is directly passed in,
+      // it will not properly get reset to the start of the stream if a chunk upload needs to be retried
+      let openUploadStream: () => NodeJS.ReadableStream
 
       if (totalFileSize < buffer.byteLength) {
         // compression did not help with reducing the size, use a readable stream from the original file for upload
-        uploadStream = fs.createReadStream(parameters.file)
+        openUploadStream = () => fs.createReadStream(parameters.file)
         isGzip = false
         uploadFileSize = totalFileSize
       } else {
         // create a readable stream using a PassThrough stream that is both readable and writable
-        const passThrough = new stream.PassThrough()
-        passThrough.end(buffer)
-        uploadStream = passThrough
+        openUploadStream = () => {
+          const passThrough = new stream.PassThrough()
```

Comment:
I realize this was the existing code, but why do we need a stream that is both readable and writable for upload?

Reply:
It's an implementation detail, we have an in-memory buffer and we have to convert somehow to a readable stream. A passthrough stream is used to get the job done without any extra pipping or secondary streams. During the actual upload, it's treated only as a readable stream so it makes no difference.

I think this is where I originally got the idea from: https://stackoverflow.com/questions/16038705/how-to-wrap-a-buffer-as-a-stream2-readable-stream

I recall experimenting with a bunch of other techniques but they all ended up being considerably more complex.

Reply:
Could we document that in https://github.com/actions/toolkit/blob/master/packages/artifact/docs/implementation-details.md?

Reply:
Done: https://github.com/actions/toolkit/blob/296a75104cc117a87b80bab8a03b7f22450b6ba4/packages/artifact/docs/implementation-details.md#uploadcompression-flow